core/templates/dev/head/pages/exploration_player/PlaythroughService.js

```diff
@@ -0,0 +1,226 @@
+  'ISSUE_TYPE_MULTIPLE_INCORRECT_SUBMISSIONS',
+  'ISSUE_TYPE_CYCLIC_STATE_TRANSITIONS', 'EARLY_QUIT_THRESHOLD_IN_SECS',
+  'NUM_INCORRECT_ANSWERS_THRESHOLD', 'NUM_REPEATED_CYCLES_THRESHOLD',
+  'CURRENT_ACTION_SCHEMA_VERSION', 'ACTION_TYPE_EXPLORATION_START',
+  'ACTION_TYPE_ANSWER_SUBMIT', 'ACTION_TYPE_EXPLORATION_QUIT',
+  function(
+      $http, PAGE_CONTEXT, PlaythroughObjectFactory, LearnerActionObjectFactory,
+      STORE_PLAYTHROUGH_URL, UrlInterpolationService, StopwatchObjectFactory,
+      CURRENT_ISSUE_SCHEMA_VERSION, ISSUE_TYPE_EARLY_QUIT,
+      ISSUE_TYPE_MULTIPLE_INCORRECT_SUBMISSIONS,
+      ISSUE_TYPE_CYCLIC_STATE_TRANSITIONS, EARLY_QUIT_THRESHOLD_IN_SECS,
+      NUM_INCORRECT_ANSWERS_THRESHOLD, NUM_REPEATED_CYCLES_THRESHOLD,
+      CURRENT_ACTION_SCHEMA_VERSION, ACTION_TYPE_EXPLORATION_START,
+      ACTION_TYPE_ANSWER_SUBMIT, ACTION_TYPE_EXPLORATION_QUIT) {
+    var playthrough = null;
+    var expStopwatch = null;
+
+    var multipleIncorrectStateName = {};
+
+    var cycleIdentifier = {};
+    var visitedStates = [];
+
+    var analyzePlaythrough = function() {
+      var timeSpentInExpInSecs = expStopwatch.getTimeInSecs();
+      // Check for multiple incorrect submissions issue.
+      if (multipleIncorrectStateName.num_times_incorrect >= (
+        NUM_INCORRECT_ANSWERS_THRESHOLD)) {
+        playthrough.issueType = ISSUE_TYPE_MULTIPLE_INCORRECT_SUBMISSIONS;
+        playthrough.issueCustomizationArgs = {
+          state_name: {
+            value: multipleIncorrectStateName.state_name
+          },
+          num_times_answered_incorrectly: {
+            value: multipleIncorrectStateName.num_times_incorrect
+          }
+        };
+        return;
+      }
+
+      // Check for cyclic state transitions issue.
+      if (cycleIdentifier.num_cycles >= NUM_REPEATED_CYCLES_THRESHOLD) {
+        playthrough.issueType = ISSUE_TYPE_CYCLIC_STATE_TRANSITIONS;
+        playthrough.issueCustomizationArgs = {
+          state_names: {
+            value: cycleIdentifier.cycle.split(',')
+          }
+        };
+        return;
+      }
+
+      // Check for early quit issue.
+      if (timeSpentInExpInSecs < EARLY_QUIT_THRESHOLD_IN_SECS) {
+        playthrough.issueType = ISSUE_TYPE_EARLY_QUIT;
+        playthrough.issueCustomizationArgs = {
+          state_name: {
+            value:
+              playthrough.actions[
+                playthrough.actions.length - 1].actionCustomizationArgs
+                .state_name.value
+          },
+          time_spent_in_exp_in_secs: {
+            value: expStopwatch.getTimeInSecs()
+          }
+        };
+        return;
+      }
+    };
+
+    var getFullPlaythroughUrl = function() {
+      return UrlInterpolationService.interpolateUrl(
+        STORE_PLAYTHROUGH_URL, {
+          exploration_id: playthrough.expId
+        });
+    };
+
+    return {
+      initSession: function(newExplorationId, newExplorationVersion) {
+        playthrough = new PlaythroughObjectFactory(
```

Comment:
nit: can you use fromBackendDict here instead? This call looks unintuitive.

Reply:
@brianrodri This is not from backend dict though. Why would we use that? These values are passed in from the PlayerServices and are initialised.

Reply:
Done.